src/coreclr/vm/comdelegate.cpp

```diff
@@ -791,6 +942,50 @@ LoaderHeap *DelegateEEClass::GetStubHeap()
     return GetInvokeMethod()->GetLoaderAllocator()->GetStubHeap();
 }
 
+#if defined(TARGET_RISCV64)
+static Stub* CreateILDelegateShuffleThunk(MethodDesc* pDelegateMD, MethodDesc* pTargetMD)
+{
+    // TODO: Stack growing is a rare case but if thunk duplication turns out to be an issue, optimize by cooking
+    // a stripped signature similar to GenerateShuffleArray and use call target from DelegateObject::GetMethodPtrAux()
+    _ASSERTE(pDelegateMD->SizeOfArgStack() < pTargetMD->SizeOfArgStack()); // IL thunk handle stack growing only
+    SigTypeContext typeContext(pDelegateMD);
+    MetaSig delegateSig(pDelegateMD);
+    INDEBUG(MetaSig targetSig(pTargetMD));
+    LOG((LF_STUBS, LL_INFO1000000,
+        "CreateILDelegateShuffleThunk %s.%s (%i args) -> %s.%s (%i args + %i this)\n",
+        pDelegateMD->GetMethodTable()->GetDebugClassName(), pDelegateMD->GetName(), delegateSig.NumFixedArgs(),
+        pTargetMD->GetMethodTable()->GetDebugClassName(), pTargetMD->GetName(), targetSig.NumFixedArgs(), !!targetSig.HasThis()
+    ));
+    _ASSERTE(delegateSig.NumFixedArgs() == targetSig.NumFixedArgs() + (targetSig.HasThis() ? 1 : 0));
+    _ASSERTE(delegateSig.GetReturnType() == targetSig.GetReturnType());
+    _ASSERTE(delegateSig.HasThis());
+
+    ILStubLinkerFlags flags = ILSTUB_LINKER_FLAG_STUB_HAS_THIS;
+    ILStubLinker stubLinker(pDelegateMD->GetModule(), pDelegateMD->GetSignature(), &typeContext, pDelegateMD, flags);
+    ILCodeStream *pCode = stubLinker.NewCodeStream(ILStubLinker::kDispatch);
+
+    for (unsigned i = 0; i < delegateSig.NumFixedArgs(); ++i)
+        pCode->EmitLDARG(i);
+
+    pCode->EmitCALL(pCode->GetToken(pTargetMD), delegateSig.NumFixedArgs(), delegateSig.IsReturnTypeVoid() ? 0 : 1);
+    pCode->EmitRET();
+
+    PCCOR_SIGNATURE pSig;
+    DWORD cbSig;
+    pDelegateMD->GetSig(&pSig, &cbSig);
+    PTR_Module pLoaderModule = pDelegateMD->GetLoaderModule();
+    MethodDesc* pStubMD = ILStubCache::CreateAndLinkNewILStubMethodDesc(
+        pDelegateMD->GetLoaderAllocator(),
+        pDelegateMD->GetMethodTable(),
+        ILSTUB_MULTICASTDELEGATE_INVOKE,
```

Comment:
Overloading`ILSTUB_MULTICASTDELEGATE_INVOKE` for shuffle thunks does not look right. Can it lead to problems?

Reply:
Don't know if it can lead to problems, but yes multicast delegates are a different use case. Maybe introduce another flag like ILSTUB_DELEGATE_SHUFFLETHUNK_STUB ?

Reply:
Yes, if you mark this as multicast delegate stub, it will confuse codepaths that check for multicast delegate stub and expect to behave in certain way, like this one: https://github.com/dotnet/runtime/blob/f1bcbeb5fa2fe84698b62d88dd35199f0d7fbedb/src/coreclr/vm/stubmgr.cpp#L1273-L1281 .

Reply:
Right, added new stub type.